src/Symfony/Component/DependencyInjection/Loader/Config/ServicesConfig.php

```diff
@@ -0,0 +1,141 @@
+ *   autoconfigure?: bool,
+ *   bind?: array<string, mixed>,
+ *   constructor?: string,
+ *   from_callable?: mixed,
+ * }
+ * @phpstan-type Alias string|array{
+ *   alias: string,
+ *   public?: bool,
+ *   deprecated?: Deprecation,
+ * }
+ * @phpstan-type Prototype array{
+ *   resource: string,
+ *   namespace?: string,
+ *   exclude?: string|list<string>,
+ *   parent?: string,
+ *   shared?: bool,
+ *   lazy?: bool|string,
+ *   public?: bool,
+ *   abstract?: bool,
+ *   deprecated?: Deprecation,
+ *   factory?: Callback,
+ *   arguments?: Arguments,
+ *   properties?: array<string, mixed>,
+ *   configurator?: Callback,
+ *   calls?: list<Call>,
+ *   tags?: Tags,
+ *   resource_tags?: Tags,
+ *   autowire?: bool,
+ *   autoconfigure?: bool,
+ *   bind?: array<string, mixed>,
+ *   constructor?: string,
+ * }
+ * @phpstan-type Stack array{
+ *   stack: list<Definition|Alias|Prototype|array<class-string, Arguments|null>>,
+ *   public?: bool,
+ *   deprecated?: Deprecation,
+ * }
+ * @phpstan-type Arguments list<mixed>|array<string, mixed>
+ * @phpstan-type Callback string|array{0:string|Reference|ReferenceConfigurator,1:string}|\Closure|Reference|ReferenceConfigurator|Expression
+ * @phpstan-type Tags list<string|array<string, array<string, mixed>>>
+ * @phpstan-type Deprecation array{package: string, version: string, message?: string}
+ * @phpstan-type Call array<string, Arguments>|array{0:string, 1?:Arguments, 2?:bool}|array{method:string, arguments?:Arguments, returns_clone?:bool}
+ */
+class ServicesConfig implements ConfigBuilderInterface
+{
+    /**
+     * $services:
+     * @param array<string, Definition|Alias|Prototype|Stack>|array<class-string, Arguments|null> $services
+     *
+     * $defaults:
+     * @param array{
+     *   public?: bool,
+     *   tags?: Tags,
+     *   resource_tags?: Tags,
+     *   autowire?: bool,
+     *   autoconfigure?: bool,
+     *   bind?: array<string, mixed>,
+     * } $defaults
+     *
+     * $instanceof:
+     * @param array{
+     *   shared?: bool,
+     *   lazy?: bool|string,
+     *   public?: bool,
+     *   properties?: array<string, mixed>,
+     *   configurator?: Callback,
+     *   calls?: list<Call>,
+     *   tags?: Tags,
+     *   resource_tags?: Tags,
+     *   autowire?: bool,
+     *   bind?: array<string, mixed>,
+     *   constructor?: string,
+     * } $instanceof
+     */
+    public function __construct(
+        private array $services,
+        array $defaults = [
+            'autowire' => true,
+            'autoconfigure' => true,
+        ],
+        array $instanceof = [],
+    ) {
+        if (isset($services['_defaults']) || isset($services['_instanceof'])) {
+            throw new InvalidArgumentException('The $services argument should not contain "_defaults" or "_instanceof" keys, use the $defaults and $instanceof parameters instead.');
+        }
+        $this->services['_defaults'] = $defaults += ['autowire' => true, 'autoconfigure' => true];
```

Comment:
Note this line: services configured using the proposed `ServicesConfig` would default to being autowired and autoconfigured.

This means a future version of recipes that defaults to PHP arrays config could start with:

```php
<?php

use Symfony\Component\DependencyInjection\Loader\Config\ParametersConfig;
use Symfony\Component\DependencyInjection\Loader\Config\ServicesConfig;

return [
    new ParametersConfig([
    ]),
    new ServicesConfig([
        'App\\' => ['resource' => '../src/']
    ]);
];
```

That's a DX proposal.

Reply:
I reverted this part, we could provide pre-configured defaults in a separate class if we'd want to.